### ISSUE TITLE
fatal error - ccalloc would have returned NULL

### ISSUE THREAD

**Issue by masrukajannat466** [wbhart/flint2#1159]
As disussed [here](https://github.com/wbhart/antic/issues/78) opening a ticket for ccalloc error

`test_flint.cpp` has same code as [crt](https://github.com/wbhart/flint2/blob/trunk/examples/crt.cpp)

```
g++ test_flint.cpp -L/mingw64/lib -lgmp -lflint
./a.exe

```
output
```

      0 [main] a (6660) child_copy: cygheap read copy failed, 0x18034B408..0x180373A68, done 0, windows pid 6660, Win32 error 6
    249 [main] a (6660) C:\msys64\mingw64\a.exe: *** fatal error - ccalloc would have returned NULL

```

My system Details:
I am running on MinGW64 with host OS being windows 10.
MinGW64 was installed as part of msys2 which is installed via exe from https://www.msys2.org

**Comment by masrukajannat466** [comment]
While setting up flint I had warnings like these which may be related

```
mul_heap_threaded.c:529:13: warning: '_nmod_mpoly_mul_heap_part1' accessing 48 bytes in a region of size 8 [-Wstringop-overflow=]
```

![flint_warnings](https://user-images.githubusercontent.com/107162465/172975767-b5b5ccda-f621-48d6-ba91-62d4e35ba7cd.png)

@wbhart  You were talking about configuring "exception mechanism" while setting up MinGW64 . Maybe it's required, maybe not, I am not sure m but I couldn't find any such option . Still exploring.

**Comment by albinahlback** [comment]
I believe that the C++ interface hasn't been maintained for years, and I would not expect it to be fixed soon.

**Comment by albinahlback** [comment]
We are not sure about how to interpret these warnings, but is currently known to only happen for GCC 11.2 and up I believe. We have an issue for this already. However, these are not related to your errors.

**Comment by wbhart** [comment]
Thanks for opening the ticket @masrukajannat466 

@tthsqe12 and I started to look into this today, and I will continue to look into it early next week. I'll post here when I have some idea what the issue is.

Obviously it is running out of memory somewhere, but I don't know what is causing that just yet.

**Comment by wbhart** [comment]
I was unable to replicate this issue with MinGW2. Here is what I did to try and replicate the issue:

```
* go to msys2.org
* download installer, follow instructions 1-5
* run MSYS2 MinGW x64 from Start menu
* follow instruction 7, installing all (the default)
* pacman -S git
* wget https://ftp.gnu.org/gnu/gmp/gmp-6.2.1.tar.bz2
* tar -xvf gmp-6.2.1.tar.bz2
* cd gmp-6.2.1
* ./configure --enable-shared --disable-static
* make -j install
* cd ..
* wget https::/www.mpfr.org/mpfr_current/mpfr-4.1.0.tar.bz2
* tar -xvf mpfr-4.1.0.tar.bz2
* cd mpfr-4.1.0
* ./configure --with-gmp=/mingw64
* make -j install
* cd ..
* git clone https://github.com/wbhart/flint2
* cd flint2
* ./configure --enable-shared --disable-static --with-gmp=/mingw64 --with-mpfr=/mingw64 --prefix=/mingw64
* make -j install
* cp libflint* /mingw64/bin
* make -j check
* cd examples
* g++ crt.cpp -o crt.exe -L/mingw64/bin -I/mingw64/include/flint -lgmp -lflint -lpthread
* ./crt 2342
```

The last step should not actually be necessary as Flint should put its dlls in the right place assuming bin is now the place MinGW now expects to find dlls. But this is something we can fix.

**Comment by wbhart** [comment]
I've been looking into what is required to fix the MinGW64 build. I've been referring to [1] which seems sane.

It seems that the dll should go in the 'mingw64/bin' directory. The import library 'libflint-17.a' should go in '/mingw/lib' so applications know where to find the dll functions (it's not clear if it helps them find the dll). Currently we don't seem to be making an import library on MinGW64.

This is done by passing the following to gcc when building the dll:

```
-Wl,--subsystem,windows,--out-implib,libflint-17.a
```

(The 'subsystem-windows' is not needed but is a convention on Windows.)

Also note that we currently produce three files on MinGW64:

```
libflint-17.dll
libflint.dll
libflint.dll.17
```

The last is surely useless, and I don't know if the second is necessary. I'll leave these alone for now unless someone has any ideas, but I propose to close this ticket by fixing the creation of import libraries above and putting the dlls in '/mingw64/bin' and the import libraries in '/mingw64/lib' on MinGW64.

Bizarrely, '$(DESTDIR)' seems to be empty on MinGW64. The 'install' target in our Makefile installs the dll in '$(DESTDIR)/$(PREFIX)' which seems odd. I'm not sure why the prefix is *after* the destination directory. I can't see what the destination directory is for in that case. If anyone knows, let me know. I'm certainly not going to change it, since that'll break something for sure.

In fact, at present 'DESTDIR' does not occur in 'configure' anywhere and it seems to be used in our Makefile without definition. So I am not really sure what it is for?

In configure '$(PREFIX)' is only ever set to '/usr/local' and never modified. This seems like it should be changed to '/mingw64' on MinGW64 if it was not altered by the user. Does anyone disagree?

Then in the install target we should install the library (.dll) to '$(PREFIX)/bin' and the import library (.a) to $(PREFIX)/lib' on MinGW64. Any disagreements?

[1] https://www.transmissionzero.co.uk/computing/building-dlls-with-mingw/

**Comment by masrukajannat466** [comment]
But mind you the warnings that I spoke of earlier did appear this time around too. Did you not encounter them ?


Thanks once again for putting in the effort to troubleshoot.
These steps should go into setting up documentation. 

You have a small typo here , it should be `wget https://www.mpfr.org/mpfr-current/mpfr-4.1.0.tar.bz2` instead of 
> wget https::/www.mpfr.org/mpfr_current/mpfr-4.1.0.tar.bz2

**Comment by wbhart** [comment]
Please, we have explained already that the warnings are bugs in GCC. We cannot do anything about these.

**Comment by wbhart** [comment]
There's no point putting the instructions for MinGW installation in our documentation. The system has changed every time someone tried to use it, so the instructions are by no means reliable.

**Comment by masrukajannat466** [comment]
Cool no issues. I am closing this ticket then.
Now I am stuck with mpir. Please look into [it](https://github.com/wbhart/mpir/issues/167#issuecomment-1156410394)

**Comment by wbhart** [comment]
MPIR is community maintained. If the community doesn't provide a pull request to fix the issue then I guess you are out of luck.

I suggest you use GMP if MPIR doesn't work for you.

**Comment by wbhart** [comment]
You misread it. Note the word 'or'.

**Comment by masrukajannat466** [comment]
arb installed fine.
But running fails to locate flint , what am I missing?

```
$ cd examples
$ gcc bernoulli.c -o berno.exe -L/mingw64/bin -lmpfr -lgmp -lflint -larb -lpthread
$ ./berno.exe
C:/msys64/home/hp/arb/examples/berno.exe: error while loading shared libraries: libflint-17.dll: cannot open shared object file: No such file or directory

```
arb doesnt create a subdirectory for includes, so `-I` wont be necessary.

**Comment by masrukajannat466** [comment]
Ah solved it . We need `cp libarb* /mingw64/bin` step for arb too. It works now.
Thanks

**Comment by wbhart** [comment]
Yeah @fredrik-johansson will have to apply the same fixes to the arb build system as I will be applying to Flint to fix this.

So far I didn't get the new configure and Makefile changes working yet though.
